iocore/net/quic/QUICLogFrame.h

```diff
@@ -0,0 +1,286 @@
+#pragma once
+
+#include <memory>
+#include <yaml-cpp/yaml.h>
+
+#include "QUICFrame.h"
+
+namespace QLog
+{
+class QLogFrame
+{
+public:
+  QLogFrame(QUICFrameType type) : _type(type) {}
+  virtual ~QLogFrame() {}
+
+  QUICFrameType
+  type() const
+  {
+    return this->_type;
+  }
+
+  // encode frame into YAML stype
+  virtual void encode(YAML::Node &node) = 0;
+
+protected:
+  QUICFrameType _type = QUICFrameType::UNKNOWN;
+};
+
+using QLogFrameUPtr = std::unique_ptr<QLogFrame>;
+
+//
+// convert QUICFrame to QLogFrame
+//
+class QLogFrameFactory
+{
+public:
+  // create QLogFrame
+  static QLogFrameUPtr create(QUICFrame &frame);
+};
+
+namespace Frame
+{
+  struct AckFrame : public QLogFrame {
+    AckFrame(QUICAckFrame &frame) : QLogFrame(frame.type())
```

Comment:
Can it use QUICFrame ? Now, we have QUICFrame QUICFrameInfo, QLogFrame these three are implementation for the same thing .

Reply:
There are overlaps but I'd keep those separated. I think current design is not bad.

QUICFrame and its sub classes are complete representations of each frames, which have all the details. These (especially Stream frame) are too heavy to keep long time. This is the reason we have QUICFrameInfo that contains minimal necessary information.

For QLogFrame, we do need to keep most information, but keeping Stream data is still too much. Having intermediate representation requires two copies (QUICFrame -> QLogFrame and QLogFrame -> YAML::Node). I'm not sure how much creating YAML::Node cost, but maybe we can do something like this:

```
class QLogFrameNode : public YAML::Node {
public:
  QLogFrame(const QUICAckFrame &);
  QLogFrame(const QUICStreamFrame &);
  QLogFrame(const QUICAckFrame &);
};
```

Reply:
I changed my mind a little. I read the qlog spec and found that the spec have `raw` field that seems like to keep entire frame data (the both header and payload). So we might want a way to keep all the data for people who want it.

In that case, using QUICFrame may make sense. However, the behavior should be configurable if we support it. I think we don't want such huge logs in most cases (we've never needed it for H2).

So, for the first step, I think current design is good enough. qlog spec is still a draft yet, and I don't think we should spend much time to support the minor use case. Supporting other events would be more valuable.